packages/experience-builder-sdk/src/hooks/useDetectEditorMode.tsx

```diff
@@ -0,0 +1,57 @@
+import { useEffect, useRef, useState } from 'react';
+import {
+  doesMismatchMessageSchema,
+  sendMessage,
+  tryParseMessage,
+} from '@contentful/experience-builder-core';
+import { INCOMING_EVENTS, OUTGOING_EVENTS } from '@contentful/experience-builder-core/constants';
+
+export const useDetectEditorMode = () => {
+  //Assume we are in editor mode initially if we are in iframe
+  const [isEditorMode, setIsEditorMode] = useState(inIframe());
+  const receivedMessage = useRef(false);
+
+  useEffect(() => {
+    const onMessage = (event: MessageEvent) => {
+      if (doesMismatchMessageSchema(event)) {
+        return;
+      }
+      const eventData = tryParseMessage(event);
+
+      if (eventData.eventType === INCOMING_EVENTS.RequestEditorMode) {
+        setIsEditorMode(true);
+        receivedMessage.current = true;
+        if (typeof window !== 'undefined') {
+          //Once we definitely know that we are in editor mode, we set this flag so future postMessage connect calls are not made
+          window.__EB__.isEditorMode = true;
+          window.removeEventListener('message', onMessage);
+        }
+      }
+    };
+
+    //Double check if we are in editor mode by listening to postMessage events
+    if (typeof window !== 'undefined' && !window.__EB__?.isEditorMode) {
+      window.addEventListener('message', onMessage);
+      sendMessage(OUTGOING_EVENTS.Connected);
+
+      setTimeout(() => {
+        if (!receivedMessage.current) {
+          // if message is not received back in time, set editorMode back to false
+          setIsEditorMode(false);
+        }
+      }, 100);
```

Comment:
I'm not sure if 100 ms will be that reliable. The web app can be quite slow for some spaces and browsers.
What would currently happen if it takes longer? I guess we show an empty screen for those 100ms and then try to fetch preview data before we finally go into editor mode, right?

**Suggestion for a follow-up ticket:**
I guess this would become relevant for customers who render the preview inside an iframe. In this case, we render an empty screen for 100ms before correctly switching to the preview. Having this 100ms pause or making it even longer doesn't sound like a good option either. Maybe we can make the initial check a bit smarter to avoid this timeout right away, e.g. we could add a query parameter in the preview URL like `cfEditor=true` so this logic here can check for it and doesn't need to rely on `inIframe` which is not a sufficient condition to assume an editor mode.

Reply:
We should test if this delay introduces an un-avoidable blink for 100ms duration. We'd want to avoid side-effects like this

Reply:
This code in the old branch wasn't really doing anything as the isInIframe check was setting editorMode to true and it wouldn't be set back to false if this check didn't go through.

The only time this would be a potential issue is outside of editor, if the user is running their own experience in their own iframe. The current code would always think it would be editor mode and thus we wouldn't support users running apps in their own iframes. This is a failsafe that flips editormode back to off if we falsely identify it in the isInIFrame check.

In development, I was seeing this callback happening in under 5ms, so I felt giving it 100ms was a good balance between being safe and not delaying the users app (when only running in their own iFrame) too long.

Reply:
Gotcha, you're right that we don't explicitly support this scenario. Also, we probably don't have specific public documentation on this. If there is a place for this in the helper docs, let's add that the preview of experiences is not supported to render inside iframes.